### PR TITLE
Added spring boot example application that can be used during test ef…

### PR DESCRIPTION
**Pull request by jvitrifork**
…forts

Added use of spring profiles

### CODE REVIEW

**Comment by jamesagnew** [comment]
I will admit I'm nervous about adding any more Boot stuff directly into the HAPI repos. The Boot stuff that is in the main repo was an outside contribution that unfortunately has been basically abandoned by its author and because none of us use boot it just ends up being a bit of a support and maintenance headache.

@jvitrifork Would you be willing to write documentation for how to use this, and commit to maintaining/supporting it long term?

**Comment by jvitrifork** [comment]
1) The intention is yes - I'm willing to do so. However - it should be clearly stated that the goal so far has been to only add this as an example - which is also why the dependencies are of scope test - the same goes for the Spring Boot Application - it is only currently added under src/test. Also, as can be seen on the PR, only the Application.java breaks if you remove all the Spring Dependencies that I've added as part of this PR - the rest is pure Spring Core - so I've actually added very little Spring Boot.

2) Preferably the whole config part should be completely refactored to use eg YAML. The current setup is error prone parsing the YAML files and the java properties file should preferably be more based on annotations. Also, lists/sets which is relevant with loading multiple IG's is a PITA in property files, where eg. YAML is more convenient.

**Comment by jvitrifork** [comment]
@jamesagnew does that address your nervousness properly?

**Comment by jvitrifork** [comment]
@seanmcilvenna can you weigh in on this feature if James is busy on other matters?

**Comment by seanmcilvenna** [comment]
I'd entertain an update to this PR that
1. includes README documentation per @jamesagnew's comment
2. includes updating the HapiProperties system to use YML. this part relates to #127

**Comment by seanmcilvenna** [comment]
And also

3. updates to resolve conflicts in the PR

**Comment by jvitrifork** [comment]
@seanmcilvenna / @jamesagnew - ok, so I've basically have rewritten almost all parts to be more "spring" conformant now. The only part that has to do with spring boot is isolated in the main class. I have also rewritten the entire configuration part. Please have a look. I works both as a standalone app doing 'mvn clean package spring-boot:repackage && java -jar target/hapi-fhir-jpaserver.war' as well as can be deployed to a web app server.

File: pom.xml [hunk 1]
```diff
@@ -301,9 +363,9 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.1</version>
                 <configuration>
-                    <source>1.8</source>
-                    <target>1.8</target>
+                    <release>11</release>
```
Comment (jamesagnew) [inline pom.xml:368]: Are we using any 11-specific features? I know there are people who deploy this project in environments where they are not currently able to use Java beyond 8 (unfortunately..)

File: src/main/resources/hapi.properties [hunk 2]
```diff
@@ -11,7 +11,7 @@ fhir_version=R4
 # (the web UI similar to the one at http://hapi.fhir.org) will use to
 # connect internally to the FHIR server, so this also needs to be a name
 # accessible from the server itself.
-server_address=http://localhost:8080/hapi-fhir-jpaserver/fhir/
+server_address=http://localhost:8080/fhir
```
Comment (jamesagnew) [inline src/main/resources/hapi.properties:14]: Could we move this property to the yaml file and then just delete this file? Seems confusing to have config split across the two

Comment (jvitrifork) [reply src/main/resources/hapi.properties:14]: hapi.properties will be deleted as soon as I have ported the tests

File: pom.xml (hunk 1 above)
Comment (jvitrifork) [reply pom.xml]: I've used Guava instead where I needed default values for sets and lists. It shoudl be Java 8 ready again. (I guess it is needless to suggest alternatives such as docker and referring to https://www.oracle.com/java/technologies/java-se-support-roadmap.html#:~:text=Java%20SE%208%20has%20gone,other%20Users%20via%20java.com.)

File: src/main/resources/hapi.properties (hunk 2 above)
Comment (jvitrifork) [reply src/main/resources/hapi.properties]: Hi @jamesagnew - I should be done now - however, there's still an issue in the tests. Would you mind having a look at whether I've wired everything up correctly in regards to the web socket subscriptions. It seems like the server does not emit events.

**Comment by oliveregger** [comment]
i just tried out the PR

```
mvn clean package spring-boot:repackage -Pboot
java -jar target/hapi-fhir-jpaserver.war
````
and implementation guide integration is now very elegant, would love to see this PR accepted!

**Comment by jamesagnew** [comment]
Hmm, I don't know why CI isn't picking up this PR... weird...

I get quite a few test failures though. Are you seeing this too @jvitrifork ?

```
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running ca.uhn.fhir.jpa.starter.ExampleServerDstu3IT
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 44.595 s <<< FAILURE! - in ca.uhn.fhir.jpa.starter.ExampleServerDstu3IT
[ERROR] ca.uhn.fhir.jpa.starter.ExampleServerDstu3IT.testCreateAndRead  Time elapsed: 0.01 s  <<< ERROR!
java.lang.IllegalStateException: Failed to load ApplicationContext
Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: javax.servlet.ServletException: jpaRestfulServer@c03579b1==ca.uhn.fhir.jpa.starter.JpaRestfulServer,jsp=null,order=1,inst=true,async=true
Caused by: ca.uhn.fhir.jpa.packages.ImplementationGuideInstallationException: Cannot install implementation guide: FHIR versions mismatch (expected <=3.0.2, package uses 4.0.1)

[ERROR] ca.uhn.fhir.jpa.starter.ExampleServerDstu3IT.testWebsocketSubscription  Time elapsed: 0 s  <<< ERROR!
java.lang.IllegalStateException: Failed to load ApplicationContext
Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: javax.servlet.ServletException: jpaRestfulServer@c03579b1==ca.uhn.fhir.jpa.starter.JpaRestfulServer,jsp=null,order=1,inst=true,async=true
Caused by: ca.uhn.fhir.jpa.packages.ImplementationGuideInstallationException: Cannot install implementation guide: FHIR versions mismatch (expected <=3.0.2, package uses 4.0.1)

[INFO] Running ca.uhn.fhir.jpa.starter.ExampleServerDstu2IT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.995 s - in ca.uhn.fhir.jpa.starter.ExampleServerDstu2IT
[INFO] Running ca.uhn.fhir.jpa.starter.MultitenantServerR4IT
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 16.877 s <<< FAILURE! - in ca.uhn.fhir.jpa.starter.MultitenantServerR4IT
[ERROR] ca.uhn.fhir.jpa.starter.MultitenantServerR4IT.testCreateAndReadInTenantA  Time elapsed: 0.008 s  <<< ERROR!
java.lang.NullPointerException
        at ca.uhn.fhir.jpa.starter.MultitenantServerR4IT.testCreateAndReadInTenantA(MultitenantServerR4IT.java:49)

[ERROR] ca.uhn.fhir.jpa.starter.MultitenantServerR4IT.testCreateAndReadInTenantB  Time elapsed: 0.009 s  <<< ERROR!
java.lang.NullPointerException
        at ca.uhn.fhir.jpa.starter.MultitenantServerR4IT.testCreateAndReadInTenantB(MultitenantServerR4IT.java:75)

[INFO] Running ca.uhn.fhir.jpa.starter.ExampleServerR4IT
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 25.338 s <<< FAILURE! - in ca.uhn.fhir.jpa.starter.ExampleServerR4IT
[ERROR] ca.uhn.fhir.jpa.starter.ExampleServerR4IT.testCreateAndRead  Time elapsed: 0 s  <<< ERROR!
java.lang.IllegalStateException: Failed to load ApplicationContext
Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: javax.servlet.ServletException: jpaRestfulServer@c03579b1==ca.uhn.fhir.jpa.starter.JpaRestfulServer,jsp=null,order=1,inst=true,async=true
Caused by: ca.uhn.fhir.jpa.packages.ImplementationGuideInstallationException: Cannot install implementation guide: FHIR versions mismatch (expected <=4.4.0, package uses 4.0.1)

[ERROR] ca.uhn.fhir.jpa.starter.ExampleServerR4IT.testWebsocketSubscription  Time elapsed: 0 s  <<< ERROR!
java.lang.IllegalStateException: Failed to load ApplicationContext
Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: javax.servlet.ServletException: jpaRestfulServer@c03579b1==ca.uhn.fhir.jpa.starter.JpaRestfulServer,jsp=null,order=1,inst=true,async=true
Caused by: ca.uhn.fhir.jpa.packages.ImplementationGuideInstallationException: Cannot install implementation guide: FHIR versions mismatch (expected <=4.4.0, package uses 4.0.1)

[INFO] Running ca.uhn.fhir.jpa.starter.ExampleServerR5IT
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 21.131 s <<< FAILURE! - in ca.uhn.fhir.jpa.starter.ExampleServerR5IT
[ERROR] ca.uhn.fhir.jpa.starter.ExampleServerR5IT.testCreateAndRead  Time elapsed: 0 s  <<< ERROR!
java.lang.IllegalStateException: Failed to load ApplicationContext
Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: javax.servlet.ServletException: jpaRestfulServer@c03579b1==ca.uhn.fhir.jpa.starter.JpaRestfulServer,jsp=null,order=1,inst=true,async=true
Caused by: ca.uhn.fhir.jpa.packages.ImplementationGuideInstallationException: Cannot install implementation guide: FHIR versions mismatch (expected <=4.4.0, package uses 4.0.1)

[ERROR] ca.uhn.fhir.jpa.starter.ExampleServerR5IT.testWebsocketSubscription  Time elapsed: 0 s  <<< ERROR!
java.lang.IllegalStateException: Failed to load ApplicationContext
Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
Caused by: javax.servlet.ServletException: jpaRestfulServer@c03579b1==ca.uhn.fhir.jpa.starter.JpaRestfulServer,jsp=null,order=1,inst=true,async=true
Caused by: ca.uhn.fhir.jpa.packages.ImplementationGuideInstallationException: Cannot install implementation guide: FHIR versions mismatch (expected <=4.4.0, package uses 4.0.1)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ExampleServerDstu3IT.testCreateAndRead » IllegalState Failed to load Applicati...
[ERROR]   ExampleServerDstu3IT.testWebsocketSubscription » IllegalState Failed to load A...
[ERROR]   ExampleServerR4IT.testCreateAndRead » IllegalState Failed to load ApplicationC...
[ERROR]   ExampleServerR4IT.testWebsocketSubscription » IllegalState Failed to load Appl...
[ERROR]   ExampleServerR5IT.testCreateAndRead » IllegalState Failed to load ApplicationC...
[ERROR]   ExampleServerR5IT.testWebsocketSubscription » IllegalState Failed to load Appl...
[ERROR]   MultitenantServerR4IT.testCreateAndReadInTenantA:49 NullPointer
[ERROR]   MultitenantServerR4IT.testCreateAndReadInTenantB:75 NullPointer
[INFO] 
[ERROR] Tests run: 9, Failures: 0, Errors: 8, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-failsafe-plugin:3.0.0-M5:verify (default) @ hapi-fhir-jpaserver-starter ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------

```

File: pom.xml (hunk 1 above)
Comment (jamesagnew) [reply pom.xml]: Yeah, as much as I'd love to move us to Java 11, it just isn't feasible for a lot of orgs around the world... Supporting a healthcare library is fun sometimes.. :)

**Comment by jamesagnew** [comment]
Well, looks like Travis is running. It's just not reporting to GitHub for some reason. I really need to migrate this project to Azure CI

https://travis-ci.org/github/hapifhir/hapi-fhir-jpaserver-starter/builds/725225700

**Comment by jkiddo** [comment]
Yep - I have those errors as well. I suspect that is is because that https://github.com/hapifhir/hapi-fhir-jpaserver-starter/blob/c9c944d236943ac9fa9e83104c7557eb3b81eef1/src/main/java/ca/uhn/fhir/jpa/starter/ApplicationContext.java#L30 has not found its way into the new structure

**Comment by jkiddo** [comment]
How is the SubscriptionTriggeringSvcImpl triggered. Is it a service that should be started somehow?

**Comment by jamesagnew** [comment]
It's in SubscriptionSubmitterConfig.java, which is a separate Spring App
Context.. Maybe it needs to be explicitly imported somewhere?

**Comment by jkiddo** [comment]
@jamesagnew - now the MultitenantServerR4IT seems to fail, as when adding the partitioningprovider, it somehow 'overshades' the regular endpoints. Not even metadata is accessible once partitioning provider is registered

**Comment by jvitrifork** [comment]
Strike that - @jamesagnew I believe theres a bug in the ResourceBinding.getMethod. As can be seen, when debugging the MultitenantServerR4IT tests, a call to $partition-management-create-partition resolves to the transaction method, which is why the tests fail, and why the server claims that it expects a bundle and not a parameter.

**Comment by jvitrifork** [comment]
wait ... it actually goes for all system wide operations. All system calls resolve to the transaction method ...

**Comment by jvitrifork** [comment]
dough ... ofc. I was missing the tenant interceptor on the client side ...

**Comment by jvitrifork** [comment]
Okay ... its down to the last test cases that regards Empi. I dont understand the test case. Why does it query for Persons when there are no Persons created? From the current Empi settings, I cant see any potential mappings between Persons and Patients. @jamesagnew could you please elaborate?

**Comment by jamesagnew** [comment]
I'm actually not sure (I didn't write that test) - Does it work in master?
Is it possible there is another Spring context that needs importing?

**Comment by jvitrifork** [comment]
It works on master, yes, so it concerns the proper registration/import of the EMPI settings only

**Comment by psavva** [comment]
Following, I would love to see this make it into JPA soon 👍

**Comment by jvitrifork** [comment]
@jamesagnew I think you need to have a look at the beans definitions for the empi package. Using component scanning, I either end up getting:


```
Description:

The bean 'empiSearchParamSvc', defined in class path resource [ca/uhn/fhir/jpa/empi/config/EmpiSubmitterConfig.class], could not be registered. A bean with that name has already been defined in URL [jar:file:/Users/jvi/.m2/repository/ca/uhn/hapi/fhir/hapi-fhir-jpaserver-empi/5.1.0/hapi-fhir-jpaserver-empi-5.1.0.jar!/ca/uhn/fhir/jpa/empi/svc/EmpiSearchParamSvc.class] and overriding is disabled.

Action:

Consider renaming one of the beans or enabling overriding by setting spring.main.allow-bean-definition-overriding=true
```

- or (with spring.main.allow-bean-definition-overriding=true)

```
Description:

Field myEmpiLinkSvc in ca.uhn.fhir.jpa.empi.svc.EmpiMatchLinkSvc required a single bean, but 2 were found:
	- empiLinkSvcImpl: defined in URL [jar:file:/Users/jvi/.m2/repository/ca/uhn/hapi/fhir/hapi-fhir-jpaserver-empi/5.1.0/hapi-fhir-jpaserver-empi-5.1.0.jar!/ca/uhn/fhir/jpa/empi/svc/EmpiLinkSvcImpl.class]
	- empiLinkSvc: defined by method 'empiLinkSvc' in class path resource [ca/uhn/fhir/jpa/empi/config/EmpiConsumerConfig.class]


Action:

Consider marking one of the beans as @Primary, updating the consumer to accept multiple beans, or using @Qualifier to identify the bean that should be consumed

```

**Comment by jamesagnew** [comment]
Pull request would gladly be accepted.

**Comment by jvitrifork** [comment]
@jamesagnew the tests runs all green now. I still would advise you guys to have a look at how and when you construct your beans as the current setup yields ambigouity, in my humble (and Spring Boot's) opinion. This project could be even more refined and tweaked even more, and made even more 'spring native', but it is intentionally (and due to scope of my own private time) now evenly split between reading configuration and construction of beans - which usually are a bit tighter than even what this PR does.

image

**Comment by jvitrifork** [comment]
@jamesagnew you need anything else before you can accept this?

**Comment by seanmcilvenna** [comment]
This looks good as far as I'm concerned. This is a big PR, though... @jamesagnew any last thoughts before I merge?

**Comment by jamesagnew** [comment]
I'm good with it.

**Pull request merged**